### PR TITLE
sign

### PR DESCRIPTION
**Pull request by tadad**

### CODE REVIEW

**Comment by dr-frmr** [comment]
I started the large task of bringing `crypto` from zuse into smart.hoon in this branch: #85

**Comment by tadad** [comment]
Changes here
* ability for wallet to sign typed messages (read arms need more work)
* `%take-with-sig` added to `fungible.hoon` - allowing for gasless approves
* `+$typed-message` added to `smart.hoon` - a standard format for signing messages. The idea is to say "by signing this data, you are letting some party modify this rice with this wheat on this town."

File: lib/zig/contracts/fungible.hoon [hunk 1]
```diff
@@ -109,7 +111,57 @@
...
+      ::  %take-with-sig allows for gasless approvals for transferring tokens
+      ::  the giver must sign the from-rice id and the typed +$approve struct above
+      ::  and the taker will pass in the signature to take the tokens
+      =/  giv=grain  (~(got by owns.cart) from-rice.args)
+      ?>  ?=(%& -.germ.giv)
+      =/  giver=account:sur  ;;(account:sur data.p.germ.giv)
+      ::  reconstruct the typed message and hash
+      =/  =typed-message
+        :-  (fry-rice holder.giv me.cart town-id.cart salt.p.germ.giv)
+        (sham ;;(approve:sur [holder.giv to.args amount.args nonce.args deadline.args]))
```
Comment (dr-frmr) [inline lib/zig/contracts/fungible.hoon:126]: you don't have to `;;` here, the sham will generate the same result regardless

File: lib/zig/contracts/fungible.hoon [hunk 2]
```diff
@@ -56,12 +56,12 @@
...
         =/  new=grain
-          [- me.cart to.args town-id.cart [%& salt.p.germ.giv [0 ~ metadata.giver]]]
+          [- me.cart to.args town-id.cart [%& salt.p.germ.giv [0 ~ metadata.giver 0]]]
         ::  continuation call: %give to rice we issued
         :+  %|
-          :+  me.cart  town-id.cart
-          [caller.inp `[%give to.args `id.new amount.args] (silt ~[id.giv]) (silt ~[id.new])]
-        [~ (malt ~[[id.new new]]) ~]
+          :~  [me.cart town-id.cart `[%give to.args `id.new amount.args] (silt ~[id.giv]) (silt ~[id.new])]
+          ==
```
Comment (dr-frmr) [inline lib/zig/contracts/fungible.hoon:63]: better to use a wideform structure here, like `~[item]` or `item^~`, to create the list

File: lib/zig/contracts/fungible.hoon [hunk 3]
```diff
@@ -109,7 +111,57 @@
...
+        (sham ;;(approve:sur [holder.giv to.args amount.args nonce.args deadline.args]))
+      =/  signed-hash  (sham typed-message)
+      
+      ::  in production this won't work because we don't have access to zuse...
+      =,  secp256k1:secp:crypto
+      ::  recoer the address from the message and signature
+      =/  recovered-address
+        %-  address-from-pub
+        %-  serialize-point
+        (ecdsa-raw-recover signed-hash sig.args)
```
Comment (dr-frmr) [inline lib/zig/contracts/fungible.hoon:135]: were you able to test this?

I added `+crypto` to smart and see you added `address-from-pub` from ethereum lib, let's see if we can get this contract to compile through +zig!deploy

File: lib/zig/contracts/fungible.hoon [hunk 4]
```diff
@@ -28,16 +28,16 @@
...
 =,  fungible
 |_  =cart
 ++  write
   |=  inp=embryo
   ^-  chick
   |^
-  ?~  args.inp  !!
-  (process ;;(arguments:sur u.args.inp) (pin caller.inp))
+  ?~  action.inp  !!
+  (process ;;(arguments:sur u.action.inp) id.from.cart)
```
Comment (dr-frmr) [inline lib/zig/contracts/fungible.hoon:40]: I made some changes to fungible.hoon structure, no logic changes just matching it with the refactoring in my branch. Probably will be an ugly merge but you can probably safely just override what's here other than nonces in `account`s and the `%take-with-sig` arm

biggest change is making use of `result`, `continuation`, `call` helpers in smart.hoon now

File: lib/zig/contracts/fungible.hoon (hunk 3 above)
Comment (tadad) [reply lib/zig/contracts/fungible.hoon:134]: right now I'm getting a
```
/gen/mk-smart/hoon:<[6 21].[6 73]>
-find.end
-find.crypto
dojo: generator failure
```
just to use the deploy generator. Tests are working fine

Comment (tadad) [reply lib/zig/contracts/fungible.hoon]: managed to solve those two, now I'm getting
```
/gen/deploy/hoon:<[22 11].[22 70]>
-find.now.cart
-find.now
dojo: generator failure
```
and if I fiddle around some more it's all cart related errors, like `-find.me.cart`

Comment (tadad) [reply lib/zig/contracts/fungible.hoon]: yeah it's weird. Currently working on all the weird merge conflicts

Comment (tadad) [reply lib/zig/contracts/fungible.hoon]: yeah I'm actually getting the same errors when I just checkout `dr/contract-qol-upgrades` and run it with no changes

**Pull request closed without merging**